### ISSUE TITLE
Unable to regain keyboard focus after sleep/lock

### ISSUE THREAD

**Issue by qwertymodo** [devinacker/bsnes-plus#52]
If my computer goes to sleep or I lock it while bsnes-plus is running, when I return, it refuses to accept any input unless I completely close the program and reopen.  It doesn't crash, emulation resumes normally when you click on the window, but you can't press any keys.

**Comment by devinacker** [comment]
Interesting. I assume this is under Windows?

**Comment by qwertymodo** [comment]
Sorry, yes. Windows 7 x64.

**Comment by mrehkopf** [comment]
Funny, I was going to report a similar issue: This also applies when disconnecting and reconnecting the mouse/keyboard while bsnes-plus is running. (I use a KVM switch on a regular basis so I tripped over this.)
All the dialogues work fine (debugger, settings, etc.) but the main emulation windows doesn't pick up any keyboard events anymore.

**Comment by qwertymodo** [comment]
Yes, I guess I wasn't very clear about the symptoms.  Input into all of the secondary windows, config dialogs, etc works fine, I can click things, the UI responds, etc.  It's just the emulation itself that stops responding to input.

**Comment by devinacker** [comment]
In both cases, does it make a difference which input driver you're using (directinput / rawinput)?

**Comment by qwertymodo** [comment]
Just tested, it happens on both input drivers.

**Comment by mrehkopf** [comment]
Good point. For me only RawInput is affected. (Changing the input driver only "took" after restarting bsnes-plus)

**Comment by devinacker** [comment]
Took a look now that I'm at home, seems to happen to me with both input drivers, so the actual issue may be somewhere else.

I don't have the same issue with higan v098, so I'll do a quick comparison in a bit.

**Comment by qwertymodo** [comment]
I just looked into this a bit more, and I actually think I know what's happening.  It's not actually losing keyboard focus, it thinks one of the modifier keys is held down when it isn't.  After locking my computer and returning to find these symptoms, I opened up the input configuration and tried binding a key, only to find that it registered Ctrl+(the key I pressed) even though Ctrl wasn't pressed at all.  Another time, same thing except with Super instead of Ctrl.  So it's still registering input just fine, it's just borking the modifier state, resulting in keypresses not being processed as expected.

**Comment by devinacker** [comment]
Interesting, good catch. I'll see what I can do about that.

**Comment by qwertymodo** [comment]
In the meantime, at least I can work around the issue by mashing all of my modifier keys until they register properly.  Before, I was just having to savestate and then close/reopen the emulator.

**Comment by devinacker** [comment]
Testing it more, rawinput and DirectX both continue to indicate that whatever modifier keys were down when locking are still down when unlocking, for some reason, so _something_ is missing key release events once the lock happens. Not really sure what or why though.
